typing/typecore.ml

```diff
@@ -6783,13 +6783,16 @@ let report_error ~loc env = function
   | Name_type_mismatch (kind, lid, tp, tpl) ->
       let type_name = Datatype_kind.type_name kind in
       let name = Datatype_kind.label_name kind in
+      let pr = match kind with
+        | Datatype_kind.Record -> quoted_longident
+        | Datatype_kind.Variant -> quoted_constr
```

Comment:
The point is that  with raw identifiers `true` is now both a valid record field and a valid constructor, but the record field must be escaped
```ocaml
type r = { \#true: int }
```
whereas the variant constructor must be left unescaped
```
type s = true
```

In practice, the variant case doesn't happen because this error path is only reachable with qualified name, and
`M.true` is a syntax error.

Reply:
I guess that the root of evil here is to pretend that the constructor `true` is an identifier; things would be simpler if reserved keywords also had reserved AST constructs.

Reply:
Another option would be to forbids `\#true` and `\#false` because according to OCaml lexing rule `true` and `false` clearly starts with a capital letter.

Reply:
I think that this would also be perfectly reasonable. I have no opinion and I think that not showing `\#true` in real-world programs is probably more important than precisely how we deal with oddball programs, so I would follow your preference.

Reply:
I am inclined to fix first the printing bugs with this PR and continue the discussion later in #13246 .